services/director-v2/tests/unit/dynamic_scheduler/test_modules_dynamic_sidecar_scheduler__core2.py

```diff
@@ -0,0 +1,273 @@
+# pylint: disable=redefined-outer-name
+
+import asyncio
+import logging
+from contextlib import asynccontextmanager
+from typing import Any, AsyncIterator
+from unittest.mock import AsyncMock, call
+
+import pytest
+from simcore_service_director_v2.modules.dynamic_sidecar.scheduler._core2 import (
+    Action,
+    ExceptionInfo,
+    Workflow,
+    WorkflowRunnerManager,
+    mark_step,
+)
+from simcore_service_director_v2.modules.dynamic_sidecar.scheduler._core2._context_base import (
+    ContextInterface,
+)
+from simcore_service_director_v2.modules.dynamic_sidecar.scheduler._core2._models import (
+    WorkflowName,
+)
+
+logger = logging.getLogger(__name__)
+
+
+# UTILS
+
+
+@asynccontextmanager
+async def workflow_runner_manager_lifecycle(
+    app: AsyncMock, context: ContextInterface, workflow: Workflow
+) -> AsyncIterator[WorkflowRunnerManager]:
+    workflow_runner_manager = WorkflowRunnerManager(
+        context=context, app=app, workflow=workflow
+    )
+    await workflow_runner_manager.setup()
+    yield workflow_runner_manager
+    await workflow_runner_manager.teardown()
+
+
+# FIXTURES
+
+
+@pytest.fixture
+def app() -> AsyncMock:
+    return AsyncMock()
+
+
+@pytest.fixture
+def workflow_name() -> WorkflowName:
+    return "test_workflow_name"
+
+
+# TESTS
+
+
+async def test_workflow_runs_in_expected_order_without_errors(
+    app: AsyncMock, context: ContextInterface, workflow_name: WorkflowName
+):
+    call_tracker = AsyncMock()
+
+    @mark_step
+    async def initialize_workflow() -> dict[str, Any]:
+        await call_tracker(initialize_workflow)
+        return {"number_a": 10, "number_b": 2}
+
+    @mark_step
+    async def compute_product(number_a: int, number_b: int) -> dict[str, Any]:
+        await call_tracker(compute_product)
+        product = number_a * number_b
+        return {"product": product}
+
+    @mark_step
+    async def check_result(product: int) -> dict[str, Any]:
+        await call_tracker(check_result)
+        assert product == 20
+        return {}
+
+    workflow = Workflow(
+        Action(
+            name="initialize",
+            steps=[
+                initialize_workflow,
+            ],
+            next_action="compute",
+            on_error_action=None,
+        ),
+        Action(
+            name="compute",
+            steps=[
+                compute_product,
+            ],
+            next_action="validate",
+            on_error_action=None,
+        ),
+        Action(
+            name="validate",
+            steps=[
+                check_result,
+            ],
+            next_action=None,
+            on_error_action=None,
+        ),
+    )
+
+    async with workflow_runner_manager_lifecycle(
+        app, context, workflow
+    ) as workflow_runner_manager:
+
+        await workflow_runner_manager.start_workflow_runner(
+            workflow_name=workflow_name, action_name="initialize"
```

Comment:
question here: why do you need to define the action_name since that is anyway the first one in your workflow? Maybe consider changing having a default that takes the first one, which is probably 95% of the cases.

Reply:
Currently you can literally add two workflows together. This is helpful if you want to switch from one say "service_start" workflow to a "service_stop" workflow.
I agre that it would be convenient to default the start to the first action, but having a default would look a bit strange.